internal/engine/task_buffer.go

```diff
@@ -0,0 +1,55 @@
+package engine
+
+import (
+	"sync"
+)
+
+// TaskBuffer represents a fixed-size buffer, it operates as a FIFO buffer
+type TaskBuffer struct {
+	tasks    []*Tasks
+	size     int
+	mu       sync.Mutex
+	notEmpty *sync.Cond
+	notFull  *sync.Cond
+}
+
+// NewTaskBuffer creates a new TaskBuffer buffer with the specified size
+func NewTaskBuffer(size int) *TaskBuffer {
+	sw := &TaskBuffer{
+		tasks: make([]*Tasks, 0, size),
+		size:  size,
+	}
+	sw.notEmpty = sync.NewCond(&sw.mu)
+	sw.notFull = sync.NewCond(&sw.mu)
+
+	return sw
+}
+
+// Add adds a new task to the task buffer, wait while the buffer is full
+func (sw *TaskBuffer) Add(task *Tasks) {
+	sw.mu.Lock()
+	defer sw.mu.Unlock()
+
+	for len(sw.tasks) >= sw.size {
+		sw.notFull.Wait()
+	}
+
+	sw.tasks = append(sw.tasks, task)
+	sw.notEmpty.Signal()
+}
+
+// Get retrieves the oldest task from the task buffer and removes it
+func (sw *TaskBuffer) Get() *Tasks {
+	sw.mu.Lock()
+	defer sw.mu.Unlock()
+
+	for len(sw.tasks) == 0 {
+		sw.notEmpty.Wait()
+	}
+
+	task := sw.tasks[0]
+	sw.tasks = sw.tasks[1:]
```

Comment:
```suggestion
The slice operation sw.tasks = sw.tasks[1:] changes the starting position of the slice, but the underlying array remains unchanged. If this operation is performed frequently (e.g., frequent task retrieval), the underlying array of the slice may grow indefinitely without releasing the used memory, potentially causing a memory leak.
So i think adding sw.tasks[0] = nil would be safer.
```

Reply:
I agree with your point, and I have made the change to address it.